robottelo/report_portal/portal.py

```diff
@@ -0,0 +1,257 @@
+        """The pretty formatter function that formats launches in a structured way
+
+        :param filter launches: Satellite or Upgrade Type launches
+
+        :returns dict: Launches, keyed with their snap_versions formatted as,
+            `{'snap_version1':launch_object1, 'snap_version2': launch_object2}`
+        """
+        formatted_launches = {}
+        for launch_info in launches:
+            launch = Launch(rp=self, launch_info=launch_info)
+            if hasattr(launch, 'satellite_version'):
+                sat_ver = launch.satellite_version
+                snap_dict = {launch.snap_version: launch}
+                if sat_ver not in formatted_launches:
+                    formatted_launches[sat_ver] = snap_dict
+                else:
+                    formatted_launches[sat_ver].update(snap_dict)
+        return formatted_launches
+
+    def _launch_requester(self):
+        """The launch GET requester to fetch the all available launches of ReportPortal
+
+        :returns dict: The json of all RP launches
+        """
+        params = {'page.page': 1, 'page.size': 500, 'page.sort': 'start_time'}
+        resp = requests.get(
+            url=f'{self.api_url}/launch', headers=self.headers, params=params, verify=False
+        )
+        resp.raise_for_status()
+        return resp.json()
+
+    def launches(self, sat_version=None, launch_type='satellite6'):
+        """Returns launches in Report Portal customized by sat_version, launch_type and
+        latest number of launches sorted by latest sat version/snap version.
+
+        This does not includes each tests data for all tests in all launches, but it includes
+        just an overview data for all and each launch in Report Portal
+
+        :param str sat_version: The satellite version
+            If its not specified, then latest count of launches of satellite versions returned
+        :param str launch_type: Either satellite6 or upgrades,
+            default returns only non-upgrade launches
+        :returns dict: The launches of Report portal.
+            if sat_version is given,
+            ```{'snap_version1':launch_object1, 'snap_version2':launch_object2}```
+            else,
+            ```{'sat_version1':{'snap_version1':launch_object1, ..}, 'sat_version2':{}}```
+        """
+        if launch_type not in launch_types:
+            raise ValueError(
+                f'Wrong Launch Type is given as {launch_type} should be one of {launch_types}'
+            )
+        data = self._launch_requester()
+        sorted_launches = sorted(
+            data['content'], key=lambda launch: launch['start_time'], reverse=True
+        )
+        typed_launches = filter(
+            lambda launch: launch['name'].lower() == launch_type, sorted_launches
+        )
+        launches_ = self._format_launches(typed_launches)
+        if sat_version:
+            if sat_version not in launches_:
+                raise ValueError(
```

Comment:
I'm not sure this should raise a ValueError, instead of just returning an empty dict and just log/print this message instead.

Reply:
Raising error because sat_version is a user input to launches def.

Lets user enter 6.7 instead of 6.7.0, so if we don't raise an error in this situation and will return only empty dict, the user might think there is no data for 6.7.
If user knows there is a data for 6.7.0, he/she might end up diagnosing what's wrong with the module :)

'sat_version' is an optional parameter and if we dont provide a value it that would give the user a chance to see all the correct satellite version names as `launches` returns all sat versions  data.